econml/solutions/causal_analysis/_causal_analysis.py

```diff
@@ -387,10 +404,36 @@ def fit(self, X, y, warm_start=False):
         # start with empty results and default shared insights
         self._results = []
         self._shared = _get_default_shared_insights_output()
+        self._shared[_CausalInsightsConstants.InitArgsKey] = {
+            'feature_inds': _sanitize(self.feature_inds),
+            'categorical': _sanitize(self.categorical),
+            'heterogeneity_inds': _sanitize(self.heterogeneity_inds),
+            'feature_names': _sanitize(self.feature_names),
+            'classification': _sanitize(self.classification),
+            'upper_bound_on_cat_expansion': _sanitize(self.upper_bound_on_cat_expansion),
+            'nuisance_models': _sanitize(self.nuisance_models),
+            'heterogeneity_model': _sanitize(self.heterogeneity_model),
+            'n_jobs': _sanitize(self.n_jobs)
+        }
 
         # convert categorical indicators to numeric indices
         categorical_inds = _get_column_indices(X, self.categorical)
 
+        # check for indices over the categorical expansion bound
+        over_bound_inds = []
+        for ind in new_inds:
+            n_cats = len(np.unique(_safe_indexing(X, ind, axis=1)))
+            if ind in categorical_inds and n_cats > self.upper_bound_on_cat_expansion:
+                warnings.warn(f"Column {ind} has more than {self.upper_bound_on_cat_expansion} values "
+                              "so no heterogeneity model will be fit for it; increase 'upper_bound_on_cat_expansion' "
+                              "to change this behavior.")
+                # can't remove in place while iterating over new_inds, so store in separate list
+                over_bound_inds.append(ind)
+        for ind in over_bound_inds:
+            new_inds.remove(ind)
+            # also remove from train_inds so we don't try to access the result later
+            train_inds.remove(ind)
+
```

Comment:
This is by design; the limit on categoricals is just to prevent excess computation/memory usage, so once those models are computed there's not much benefit to throwing them away just because the threshold was lowered, although I agree the behaviour could be a little bit confusing since the results would be different when not warm-starting.